File: python/todo_cli/todo_cli/storage.py
import json
from dataclasses import asdict
from pathlib import Path

from .models import Todo, TodoStore


DEFAULT_PATH = Path.home() / ".todo.json"


def load(path: Path = DEFAULT_PATH) -> TodoStore:
    if not path.exists():
        return TodoStore()
    raw = json.loads(path.read_text())
    todos = [Todo(**t) for t in raw["todos"]]
    return TodoStore(todos=todos, next_id=raw["next_id"])


def save(store: TodoStore, path: Path = DEFAULT_PATH) -> None:
    data = {"todos": [asdict(t) for t in store.todos], "next_id": store.next_id}
    path.write_text(json.dumps(data, indent=2))

File: elixir/todo_cli/lib/todo_cli/storage.ex
defmodule TodoCli.Storage do
  alias TodoCli.Models

  @path Path.expand("~/.todo.json")

  def load do
    case File.read(@path) do
      {:ok, content} ->
        raw = Jason.decode!(content)
        todos = Enum.map(raw["todos"], &Models.from_map/1)
        {todos, raw["next_id"]}

      {:error, :enoent} ->
        {[], 1}
    end
  end

  def save(todos, next_id) do
    data = %{"todos" => Enum.map(todos, &Models.to_map/1), "next_id" => next_id}
    File.write!(@path, Jason.encode!(data, pretty: true))
  end
end
